File: sdk-core/src/test/java/dev/restate/sdk/core/javaapi/SleepTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.core.javaapi;

import static dev.restate.sdk.core.javaapi.JavaAPITests.testDefinitionForService;

import dev.restate.sdk.DurableFuture;
import dev.restate.sdk.core.SleepTestSuite;
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder;
import dev.restate.sdk.core.TestSerdes;
import dev.restate.serde.Serde;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

public class SleepTest extends SleepTestSuite {

  @Override
  protected TestInvocationBuilder sleepGreeter() {
    return testDefinitionForService(
        "SleepGreeter",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, unused) -> {
          ctx.sleep(Duration.ofSeconds(1));
          return "Hello";
        });
  }

  @Override
  protected TestInvocationBuilder manySleeps() {
    return testDefinitionForService(
        "ManySleeps",
        Serde.VOID,
        Serde.VOID,
        (ctx, unused) -> {
          List<DurableFuture<?>> collectedDurableFutures = new ArrayList<>();

          for (int i = 0; i < 10; i++) {
            collectedDurableFutures.add(ctx.timer(Duration.ofSeconds(1)));
          }

          DurableFuture.all(
                  collectedDurableFutures.get(0),
                  collectedDurableFutures.get(1),
                  collectedDurableFutures
                      .subList(2, collectedDurableFutures.size())
                      .toArray(DurableFuture[]::new))
              .await();

          return null;
        });
  }
}

File: sdk-core/src/test/kotlin/dev/restate/sdk/core/kotlinapi/SleepTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.core.kotlinapi

import dev.restate.sdk.core.SleepTestSuite
import dev.restate.sdk.core.TestDefinitions
import dev.restate.sdk.core.kotlinapi.KotlinAPITests.Companion.testDefinitionForService
import dev.restate.sdk.kotlin.*
import kotlin.time.Duration.Companion.seconds

class SleepTest : SleepTestSuite() {

  override fun sleepGreeter(): TestDefinitions.TestInvocationBuilder =
      testDefinitionForService("SleepGreeter") { ctx, _: Unit ->
        ctx.sleep(1.seconds)
        "Hello"
      }

  override fun manySleeps(): TestDefinitions.TestInvocationBuilder =
      testDefinitionForService<Unit, Unit>("ManySleeps") { ctx, _: Unit ->
        val durableFutures = mutableListOf<DurableFuture<Unit>>()
        for (i in 0..9) {
          durableFutures.add(ctx.timer(1.seconds))
        }
        durableFutures.awaitAll()
      }
}
